pkg/k8s/init.go

```diff
@@ -67,14 +67,21 @@ func waitForNodeInformation(ctx context.Context, nodeName string) *nodeTypes.Nod
 func retrieveNodeInformation(nodeName string) (*nodeTypes.Node, error) {
 	requireIPv4CIDR := option.Config.K8sRequireIPv4PodCIDR
 	requireIPv6CIDR := option.Config.K8sRequireIPv6PodCIDR
+	// At this point it's not clear whether the device auto-detection will
+	// happen, as initKubeProxyReplacementOptions() might disable BPF NodePort.
+	// Anyway, to be on the safe side, don't give up waiting for a (Cilium)Node
+	// self object.
+	mightAutoDetectDevices :=
+		option.Config.KubeProxyReplacement != option.KubeProxyReplacementDisabled &&
+			(len(option.Config.Devices) == 0 || option.Config.DirectRoutingDevice == "")
```

Comment:
nit: Given that `GetNodeSpec()` also needs this logic it seems appropriate for a separate function. I also wonder if it would be more suitable to live near other devices logic, eg in the `option` package.